lib/core/src/Cardano/Wallet/DB/Checkpoints.hs

```diff
@@ -0,0 +1,141 @@
+module Cardano.Wallet.DB.Checkpoints
+    ( getPoint
+
+    -- * Checkpoints  
+    , Checkpoints (..)
+    , singleton
+    , getLatest
+    , findNearestPoint
+    
+    -- * Delta types
+    , DeltaCheckpoints (..)
+    , DeltaMap (..)
+    ) where
+
+import Prelude
+
+import Data.Delta
+    ( Delta (..) )
+import Data.Generics.Internal.VL.Lens
+    ( over, view )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Map.Strict
+    ( Map )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Primitive.Model as W
+import qualified Cardano.Wallet.Primitive.Types as W
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+
+{- NOTE [PointSlotNo]
+
+'SlotNo' cannot represent the genesis point.
+
+Historical hack. The DB layer can't represent 'Origin' in the database,
+instead we have mapped it to 'SlotNo 0', which is wrong.
+
+Rolling back to SlotNo 0 instead of Origin is fine for followers starting
+from genesis (which should be the majority of cases). Other, non-trivial
+rollbacks to genesis cannot occur on mainnet (genesis is years within
+stable part, and there were no rollbacks in byron).
+
+Could possibly be problematic in the beginning of a testnet without a
+byron era. /Perhaps/ this is what is happening in the
+>>> [cardano-wallet.pools-engine:Error:1293] [2020-11-24 10:02:04.00 UTC]
+>>> Couldn't store production for given block before it conflicts with
+>>> another block. Conflicting block header is:
+>>> 5bde7e7b<-[f1b35b98-4290#2008]
+errors observed in the integration tests.
+
+The issue has been partially fixed in that 'rollbackTo' now takes
+a 'SlotPoint' as argument, which can represent the 'Origin'.
+However, the database itself mostly stores slot numbers.
+
+FIXME LATER during ADP-1043: As we move towards in-memory data,
+all slot numbers in the DB file will either be replaced by
+the 'SlotPoint' type, or handled slightly differently when it
+is clear that the data cannot exist at the genesis point
+(e.g. for TxHistory).
+
+-}
+
+-- | Helper function: Get the 'Point' of a wallet state.
+getPoint :: W.Wallet s -> W.SlotPoint
+getPoint =
+    W.toSlotPoint . W.chainPointFromBlockHeader . view #currentTip
+
+{-------------------------------------------------------------------------------
+    Checkpoints
+-------------------------------------------------------------------------------}
+{- HLINT ignore Checkpoints "Use newtype instead of data" -}
+-- | Collection of checkpoints indexed by 'SlotPoint'.
+data Checkpoints a = Checkpoints
+    { checkpoints :: Map W.SlotPoint a
+    } deriving (Eq,Show,Generic)
+-- FIXME LATER during ADP-1043:
+--  Use a more sophisticated 'Checkpoints' type that stores deltas.
+
+-- | Make a single checkpoint.
+singleton :: W.SlotPoint -> a -> Checkpoints a
+singleton key a = Checkpoints $ Map.singleton key a
+
+-- | Get the checkpoint with the largest 'SlotNo'.
+getLatest :: Checkpoints a -> (W.SlotPoint, a)
+getLatest = from . Map.lookupMax . view #checkpoints 
```

Comment:
why not to use Either here with explicit Error. use could then also use `maybeToLeft` https://hackage.haskell.org/package/either-5.0.1.1/docs/src/Data.Either.Combinators.html#maybeToLeft

Reply:
Hm. This error case would be a programming error, as we assume that a checkpoint for genesis is always stored, so I thought it would be fine to use a partial function here. 🤔

But I can see that if we consider `Checkpoints` as an abstraction independent of the database, then the error might in principle occur, e.g. we have

```
getLatest (apply (RollbackTo pt2) (singleton pt1 a)) = _|_
```

I think that the appropriate solution is actually to change `singleton` to only accept the genesis slot (and rename the function to `fromGenesis`) and to comment that `RestricTo` will never remove the genesis checkpoint. Then, the `Checkpoints` data type will satisfy the invariant that the map is always nonempty.

I will add a change to this effect.

Reply:
I'm beginning to confuse myself by doing a lot rebasing, so I would like to postpone the introduction of `fromGenesis` to the next PR in this series.